sw/device/silicon_creator/lib/drivers/keymgr.c

```diff
@@ -130,3 +131,139 @@ void keymgr_advance_state(void) {
+static rom_error_t keymgr_wait_until_done(void) {
+  // Poll the OP_STATUS register until it is something other than "WIP".
+  uint32_t reg;
+  uint32_t status;
+  do {
+    // Read OP_STATUS and then clear by writing back the value we read.
+    reg = abs_mmio_read32(kBase + KEYMGR_OP_STATUS_REG_OFFSET);
+    abs_mmio_write32(kBase + KEYMGR_OP_STATUS_REG_OFFSET, reg);
+    status = bitfield_field32_read(reg, KEYMGR_OP_STATUS_STATUS_FIELD);
+  } while (status == KEYMGR_OP_STATUS_STATUS_VALUE_WIP);
+
+  // Check if the key manager reported errors. If it is already idle or
+  // completed an operation successfully, return an OK status. A `WIP` status
+  // should not be possible because of the check above.
+  switch (status) {
+    case KEYMGR_OP_STATUS_STATUS_VALUE_IDLE:
+      return kErrorOk;
+    case KEYMGR_OP_STATUS_STATUS_VALUE_DONE_SUCCESS:
+      return kErrorOk;
+    case KEYMGR_OP_STATUS_STATUS_VALUE_DONE_ERROR: {
+      // Clear the ERR_CODE register before returning.
+      uint32_t err_code = abs_mmio_read32(kBase + KEYMGR_ERR_CODE_REG_OFFSET);
+      abs_mmio_write32(kBase + KEYMGR_ERR_CODE_REG_OFFSET, err_code);
+      return kErrorKeymgrInternal;
+    }
+  }
+
+  // Should be unreachable.
+  HARDENED_TRAP();
+  return kErrorKeymgrInternal;
+}
+
+rom_error_t keymgr_generate_attestation_key_otbn(
+    keymgr_diversification_t diversification) {
+  HARDENED_RETURN_IF_ERROR(keymgr_is_idle());
+
+  // Select OTBN as the destination.
+  uint32_t ctrl =
+      bitfield_field32_write(0, KEYMGR_CONTROL_SHADOWED_DEST_SEL_FIELD,
+                             KEYMGR_CONTROL_SHADOWED_DEST_SEL_VALUE_OTBN);
+
+  // Select the attestation CDI.
+  ctrl = bitfield_bit32_write(ctrl, KEYMGR_CONTROL_SHADOWED_CDI_SEL_BIT, true);
+
+  // Select the "generate" operation.
+  ctrl = bitfield_field32_write(
+      ctrl, KEYMGR_CONTROL_SHADOWED_OPERATION_FIELD,
+      KEYMGR_CONTROL_SHADOWED_OPERATION_VALUE_GENERATE_HW_OUTPUT);
+
+  // Write the control register.
+  abs_mmio_write32_shadowed(kBase + KEYMGR_CONTROL_SHADOWED_REG_OFFSET, ctrl);
+
+  // Start the operation and wait for it to complete.
+  keymgr_start(diversification);
+  return keymgr_wait_until_done();
+}
+
+rom_error_t keymgr_sideload_clear_otbn(void) {
+  HARDENED_RETURN_IF_ERROR(keymgr_is_idle());
+
+  // Set SIDELOAD_CLEAR to begin continuously clearing the requested slot.
+  abs_mmio_write32(
+      kBase + KEYMGR_SIDELOAD_CLEAR_REG_OFFSET,
+      bitfield_field32_write(0, KEYMGR_SIDELOAD_CLEAR_VAL_FIELD,
+                             KEYMGR_SIDELOAD_CLEAR_VAL_VALUE_OTBN));
+
+  // Read back the value (hardening measure).
+  uint32_t sideload_clear =
+      abs_mmio_read32(kBase + KEYMGR_SIDELOAD_CLEAR_REG_OFFSET);
+  if (bitfield_field32_read(sideload_clear, KEYMGR_SIDELOAD_CLEAR_VAL_FIELD) !=
+      KEYMGR_SIDELOAD_CLEAR_VAL_VALUE_OTBN) {
+    return kErrorKeymgrInternal;
+  }
+
+  // Spin for 100 microseconds.
+  // TODO: this value seems to work for tests, but it would be good to run a
```

Comment:
maybe file a ROM_EXT issue (so it ends up in the ROM_EXT [board](https://github.com/orgs/lowRISC/projects/13/views/53?filterQuery=label%3ASW%3AROM_EXT+-status%3ADone&visibleFields=%5B%22Title%22%2C2972208%2C7206618%2C%22Assignees%22%2C12795604%2C%22Status%22%2C%22Milestone%22%2C16720015%2C20877873%2C11502210%2C5758667%2C2972345%5D)) to track this?